File: metrics/TikTak/examples/simple_examples.py
"""Run from the TikTak folder: python -m examples.simple_examples --workers 2."""
from __future__ import annotations

import argparse
import json
from pathlib import Path

import numpy as np

from tiktak import ModelEvaluationError, MomentObjective, TikTakConfig, minimize


def quadratic(x):
    return float(np.sum((x - np.array([0.2, -0.4]))**2))


def rastrigin(x):
    return float(10 * len(x) + np.sum(x**2 - 10 * np.cos(2 * np.pi * x)))


def income_moments(x):
    """Stationary AR(1) variance and first two autocovariances."""
    rho, sigma = x
    if abs(rho) >= 1 or sigma < 0:
        raise ModelEvaluationError("no stationary income distribution")
    variance = sigma**2 / (1 - rho**2)
    return np.array([variance, rho * variance, rho**2 * variance])


def rough_holes(x):
    if x[0] < -0.35 or (0.2 < x[0] < 0.45 and x[1] < 0.4):
        raise ModelEvaluationError("no equilibrium")
    if x[1] < -0.7:
        return np.nan
    return float(np.sum(np.round((x - np.array([0.6, 0.3])) / 0.02)**2) * 0.0004)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--workers", type=int, default=1)
    parser.add_argument("--output", type=Path, default=Path("tiktak-runs/simple"))
    args = parser.parse_args()
    target = income_moments([0.65, 0.2])
    cases = [
        ("quadratic", quadratic, [(-2, 2)] * 2, dict(n_samples=16, n_local=4), {}),
        ("rastrigin", rastrigin, [(-5.12, 5.12)] * 2,
         dict(n_samples=256, n_local=32, local_max_evals=180, seed=7), {}),
        ("ar1_moments", MomentObjective(income_moments, target, scales=target),
         [(0, 0.98), (0, np.inf)], dict(n_samples=64, n_local=8, local_max_evals=250, x_tol=1e-7),
         dict(scale=[1, 0.2])),
        ("rough_holes", rough_holes, [(-1, 1)] * 2,
         dict(n_samples=128, n_local=16, local_method="pattern", local_max_evals=150, x_tol=1e-4), {}),
    ]
    summary = {}
    for name, objective, bounds, settings, options in cases:
        config = TikTakConfig(workers=args.workers, **settings)
        result = minimize(objective, bounds, config=config, run_dir=args.output / name,
                          problem_id=f"example-{name}-v1", **options)
        summary[name] = result.to_dict()
        print(f"{name:14s} loss={result.fun:.6g}  x={result.x}  "
              f"evaluations={result.n_evals}  failures={result.n_failed}  {result.status}")
    (args.output / "summary.json").write_text(json.dumps(summary, indent=2) + "\n")


if __name__ == "__main__":
    # Required for multiprocessing's spawn mode on macOS/Linux.
    main()

File: metrics/TikTak.jl/examples/simple_examples.jl
# Small correctness checks, not evidence of performance on large structural models.
#
# Run from the TikTak.jl folder:
#     julia --project=. examples/simple_examples.jl --workers 2 [--output tiktak-runs/simple]
#
# Fits a quadratic, the multimodal Rastrigin function, an analytic AR(1) income
# process with an unbounded parameter, and a staircase objective with
# infeasible regions. Use a different --output directory to run them again.
using Distributed
using JSON

function parse_arguments(args)
    options = Dict{String,Any}("workers" => 0, "output" => "tiktak-runs/simple")
    i = 1
    while i <= length(args)
        if args[i] == "--workers"
            options["workers"] = parse(Int, args[i+1])
            i += 2
        elseif args[i] == "--output"
            options["output"] = args[i+1]
            i += 2
        else
            error("unknown argument $(args[i]); use --workers N and --output DIR")
        end
    end
    return options
end

const OPTIONS = parse_arguments(ARGS)
# Workers do not inherit the active project; pass it so `using TikTak` works there.
OPTIONS["workers"] > 0 && addprocs(OPTIONS["workers"]; exeflags="--project=$(Base.active_project())")

@everywhere using TikTak

# Objectives must be defined on every process that evaluates them.
@everywhere begin
    quadratic(x) = sum((x .- [0.2, -0.4]) .^ 2)

    rastrigin(x) = 10 * length(x) + sum(x .^ 2 .- 10 .* cos.(2π .* x))

    """Stationary AR(1) variance and first two autocovariances."""
    function income_moments(x)
        ρ, σ = x
        (abs(ρ) >= 1 || σ < 0) && throw(ModelEvaluationError("no stationary income distribution"))
        variance = σ^2 / (1 - ρ^2)
        return [variance, ρ * variance, ρ^2 * variance]
    end

    function rough_holes(x)
        if x[1] < -0.35 || (0.2 < x[1] < 0.45 && x[2] < 0.4)
            throw(ModelEvaluationError("no equilibrium"))
        end
        x[2] < -0.7 && return NaN
        return sum(round.((x .- [0.6, 0.3]) ./ 0.02) .^ 2) * 0.0004
    end
end

function main()
    target = income_moments([0.65, 0.2])
    cases = [
        ("quadratic", quadratic, [(-2, 2), (-2, 2)],
         (n_samples=16, n_local=4), (;)),
        ("rastrigin", rastrigin, [(-5.12, 5.12), (-5.12, 5.12)],
         (n_samples=256, n_local=32, local_max_evals=180, seed=7), (;)),
        ("ar1_moments", MomentObjective(income_moments, target; scales=target), [(0, 0.98), (0, Inf)],
         (n_samples=64, n_local=8, local_max_evals=250, x_tol=1e-7), (scale=[1, 0.2],)),
        ("rough_holes", rough_holes, [(-1, 1), (-1, 1)],
         (n_samples=128, n_local=16, local_method=PatternSearchLocal(), local_max_evals=150, x_tol=1e-4), (;)),
    ]
    summary = Dict{String,Any}()
    for (name, objective, bounds, settings, options) in cases
        config = TikTakConfig(; settings...)
        result = minimize(objective, bounds; config=config, run_dir=joinpath(OPTIONS["output"], name),
                          problem_id="example-$name-v1", options...)
        summary[name] = TikTak.to_dict(result)
        println(rpad(name, 14), " loss=", result.fun, "  x=", result.x, "  evaluations=", result.n_evals,
                "  failures=", result.n_failed, "  ", result.status)
    end
    open(joinpath(OPTIONS["output"], "summary.json"), "w") do io
        JSON.json(io, summary; pretty=true)
        write(io, '\n')
    end
end

main()
